presto-geospatial/src/main/java/com/facebook/presto/geospatial/BingTileFunctions.java

```diff
@@ -174,20 +175,113 @@ public static Block geometryToBingTiles(@SqlType(GEOMETRY_TYPE_NAME) Slice input
+            // of these checks. The idea is to identify large tiles which are fully covered by the
+            // geometry. For each such tile, we can cheaply compute all the containing tiles at
+            // the right zoom level and append them to results in bulk. This way we perform a single
+            // containment check instead of 2 to the power of level delta intersection checks, where
+            // level delta is the difference between the desired zoom level and level of the large
+            // tile covered by the geometry.
+            int divider = 1 << (zoomLevel - minZoomLevelForOptimizedTiling);
+            int minX = (int) Math.floor(leftUpperTile.getX() / divider);
+            int maxX = (int) Math.floor(rightLowerTile.getX() / divider);
+            int minY = (int) Math.floor(leftUpperTile.getY() / divider);
+            int maxY = (int) Math.floor(rightLowerTile.getY() / divider);
+
+            for (int x = minX; x <= maxX; x++) {
+                for (int y = minY; y <= maxY; y++) {
+                    BingTile tile = BingTile.fromCoordinates(x, y, minZoomLevelForOptimizedTiling);
+                    geometryToBingTiles(geometry, zoomLevel, tile, blockBuilder);
                 }
             }
         }
 
         return blockBuilder.build();
     }
 
+    /**
+     * Identifies a minimum set of tiles at specified zoom level that cover intersection of the
+     * specified geometry and a specified tile of the same or lower level. Adds tiles to provided
+     * BlockBuilder.
+     *
+     * @pre tile.getZoomLevel() <= zoomLevel
+     */
+    private static void geometryToBingTiles(OGCGeometry geometry,
+                                            int zoomLevel,
+                                            BingTile tile,
+                                            BlockBuilder blockBuilder)
+    {
+        int tileZoomLevel = tile.getZoomLevel();
+
+        OGCGeometry tilePolygon = tileToPolygon(tile);
+        if (tileZoomLevel == zoomLevel) {
+            if (!geometry.disjoint(tilePolygon)) {
+                BIGINT.writeLong(blockBuilder, tile.encode());
+            }
+            return;
+        }
+
+        if (geometry.contains(tilePolygon)) {
+            int n = 1 << (zoomLevel - tileZoomLevel);
+            int minX = n * tile.getX();
+            int minY = n * tile.getY();
+            for (int x = minX; x < minX + n; x++) {
+                for (int y = minY; y < minY + n; y++) {
+                    BingTile subTile = BingTile.fromCoordinates(x, y, zoomLevel);
+                    BIGINT.writeLong(blockBuilder, subTile.encode());
+                }
+            }
+            return;
+        }
+
+        if (geometry.disjoint(tilePolygon)) {
+            return;
+        }
+
+        int minX = 2 * tile.getX();
+        int minY = 2 * tile.getY();
+        int nextZoomLevel = tileZoomLevel + 1;
+        for (int i = 0; i < 2; i++) {
```

Comment:
Why don't you use `x`/`y` instead of `i`/`j` similar to other loops here? This will make it more readable below where we add `minX`/`minY` to `i`/`j`.

Reply:
Why do we use a loop condition of ` < 2` here?

Reply:
I'm moving to the next level by splitting the tile into 4.